File: src/main/java/november_2024/CheapestFlights.java
package november_2024;

import java.util.*;

public class CheapestFlights {

    public static void main(String[] args) {
        int n = 4;
//        [[0,1,1],[0,2,5],[1,2,1],[2,3,1]]
//        int[][] flights = {{0, 1, 100}, {1, 2, 100}, {0, 2, 500}};
        int[][] flights = {{0, 1, 1}, {0, 2, 5}, {1, 2, 1}, {2, 3, 1}};
        int src = 0;
        int dst = 3;
        int k = 1;

        CheapestFlights cheapestFlights = new CheapestFlights();
        int cheapestPrice = cheapestFlights.findCheapestPrice(n, flights, src, dst, k);
        System.out.println(cheapestPrice);
    }

    // O(e + k) time
    public int findCheapestPrice(int n, int[][] flights, int src, int dst, int k) {
        int[] prices = new int[n];
        Arrays.fill(prices, Integer.MAX_VALUE);
        prices[src] = 0;

        for (int i = 0; i < k + 1; i++) {
            int[] tempPrices = new int[n];
            System.arraycopy(prices, 0, tempPrices, 0, n);
            for (int[] flight : flights) {
                int s = flight[0];
                int d = flight[1];
                int p = flight[2];
                if (prices[s] == Integer.MAX_VALUE) {
                    continue;
                }
                if (prices[s] + p < tempPrices[d]) {
                    tempPrices[d] = prices[s] + p;
                }
            }
            System.arraycopy(tempPrices, 0, prices, 0, n);
        }

        return prices[dst] == Integer.MAX_VALUE ? -1 : prices[dst];
    }


}

File: python/second/cheapest_flights.py
class Solution(object):
    # O(n * k) time | O(n) space
    def findCheapestPrice(self, n, flights, src, dst, k):
        """
        :type n: int
        :type flights: List[List[int]]
        :type src: int
        :type dst: int
        :type k: int
        :rtype: int
        """
        dp = [ float("infinity") ] * n
        dp[src] = 0

        for i in range(k + 1):
            newDp = dp[::]
            for flight in flights:
                s = flight[0]
                d = flight[1]
                c = flight[2]

                if dp[s] == float("infinity"):
                    continue

                if dp[s] + c < dp[d]:
                    newDp[d] = min(newDp[d], dp[s] + c)
            dp = newDp

        return dp[dst] if dp[dst] != float("infinity") else -1
    

solution = Solution()

n = 3
flights = [[0,1,100],[1,2,100],[0,2,500]]
src = 0
dst = 2
k = 1

print(solution.findCheapestPrice(n, flights, src, dst, k))
